Support setting UnknownUrlSchemePolicy

coiby opened the pull request:
Use could set UnknownUrlSchemePolicy by `set unknown_url.scheme.policy
{1..3}`. This will fix issue #5275, i.e., a `tecent://` link will be
opened if unknown_url.scheme.policy=3.

According to https://doc.qt.io/qt-5/qwebenginesettings.html#UnknownUrlSchemePolicy-enum,
 - 1: DisallowUnknownUrlSchemes
 - 2: AllowUnknownUrlSchemesFromUserInteraction
 - 3: AllowAllUnknownUrlSchemes

This setting is only valid for webengine backend since only QWebEngineSettings provides
setUnknownUrlSchemePolicy method.

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -131,6 +138,11 @@ class WebEngineSettings(websettings.AbstractSettings):
+    _UnknownUrlSchemePolicy = {
+        'unknown_url.scheme.policy': 2,
+    }
+
```
This probably isn't needed anymore now?

The-Compiler commented on qutebrowser/config/websettings.py:
```diff
@@ -99,6 +99,7 @@ class AbstractSettings:
     _FONT_SIZES = {}  # type: typing.Dict[str, typing.Any]
     _FONT_FAMILIES = {}  # type: typing.Dict[str, typing.Any]
     _FONT_TO_QFONT = {}  # type: typing.Dict[typing.Any, QFont.StyleHint]
+    _UnknownUrlSchemePolicy = {}  # type: typing.Dict[str, typing.Any]
```
This probably isn't needed anymore now?

The-Compiler commented on qutebrowser/config/websettings.py:
```diff
@@ -177,6 +178,17 @@ def set_default_text_encoding(self, encoding: str) -> bool:
+        old_value = self._settings.unknownUrlSchemePolicy()
+        self._settings.setUnknownUrlSchemePolicy(policy)
+        return old_value != policy
+
```
This should go to `webenginesettings.py` instead, as it's QtWebEngine-specific code.

The-Compiler commented on qutebrowser/config/websettings.py:
```diff
@@ -193,6 +205,15 @@ def _update_setting(self, setting: str, value: typing.Any) -> bool:
+            try:
+                return self.set_unknown_url_scheme_policy(value)
+            except AttributeError:
+                pass
```
Might be cleaner to override the `_update_setting` method in `WebEngineSettings` - something like this (untested):

```python3
def _update_setting(self, setting: str, value: typing.Any) -> bool:
    if setting == 'unknown_url.scheme.policy':
        return self.set_unknown_url_scheme_policy(value)
    return super()._update_setting(setting, value)
```

You don't need to take care to handle Qt < 5.10 compatibility if you set the `backend` key correctly in `configdata.yml` as mentioned above, at least if I'm not mistaken.

The-Compiler commented on qutebrowser/config/websettings.py:
```diff
@@ -225,7 +246,8 @@ def update_for_url(self, url: QUrl) -> typing.Set[str]:
         for setting in (list(self._ATTRIBUTES) + list(self._FONT_SIZES) +
-                        list(self._FONT_FAMILIES)):
+                        list(self._FONT_FAMILIES) +
+                        list(self._UnknownUrlSchemePolicy)):
```
This probably isn't needed anymore now?

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -2512,6 +2512,22 @@ colors.webpage.prefers_color_scheme_dark:
+    Set UnknownUrlSchemePolicy.
+
+    Only valid for webengine backend (Qt > 5.10).
+    1: DisallowUnknownUrlSchemes, 2: AllowUnknownUrlSchemesFromUserInteraction, 3: AllowAllUnknownUrlSchemes.
```
I think there are various things which should be changed here. In general, settings are user-facing and should hide implementation details from the underlying engine(s). As-is, this setting (and its values) is pretty user-unfriendly.

Some suggestions:

- Use a string with `valid_values` set (see other settings in the file for some examples), so that the options have a name rather than being a number you need to describe in the documentation. You'll then need to map that string to the enum member inside `set_unknown_url_scheme_policy`. See the [Qt documentation](https://doc.qt.io/qt-5/qwebenginesettings.html#UnknownUrlSchemePolicy-enum) for some inspiration about the descriptions for the values.
- Use the existing facilities to note compatibility instead of having it in the text, something like:

```
  backend:
    QtWebEngine: Qt 5.10
    QtWebKit: false
```

- Have a better text than "Set UnknownUrlSchemePolicy" - something that tells an user what this setting *does* exactly. Something like "How navigation requests to URLs with unknown schemes are handled." for example.

coiby replied:
Thanks for the all the handy suggestions! Updated in the latest commit!

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -2512,6 +2512,22 @@ colors.webpage.prefers_color_scheme_dark:
 
+## unkown_url
+
+unknown_url.scheme.policy:
```
I don't think it makes sense to open a new "group" of settings with only this setting in it. What about `content.unknown_url_scheme_policy`?

coiby commented:
On Tue, Apr 14, 2020 at 06:08:27AM -0700, Florian Bruhin wrote:
>@The-Compiler requested changes on this pull request.
>
>Can you elaborate on the changes you did here? I feel like there's a lot of speculative generality (with `AbstractSettingsWrapper` and such) which isn't actually used in the end (because QtWebKit doesn't have multiple settings objects, and QtWebKit doesn't have an unknown URL scheme setting).
Yes, the solution needn't to be so complex. I put AbstractSettingsWrapper
back to webenginesettings.py in the latest commit.

>I think a much simpler implementation should be possible here, no? Something like calling `set_unknown_url_scheme_policy` from the needed places in the QtWebEngine settings only, and leaving all the rest untouched?

Catching AttributeError could deal with all the cases where webkit and
webengine<5.11 don't support setting UnknownUrlSchemePolicy. But we
need to put this part of code in qutebrowser/config/websetting.py so
setting unknown_url.scheme.policy in webkit backend won't cause any trouble.

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -2512,6 +2512,22 @@ colors.webpage.prefers_color_scheme_dark:
+## unkown_url
+
+unknown_url.scheme.policy:
+  default: 1
```
The default from Qt is `AllowUnknownUrlSchemesFromUserInteraction`. We should keep that.

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -151,6 +158,19 @@ class WebEngineSettings(websettings.AbstractSettings):
+            'AllowAllUnknownUrlSchemes':
+                QWebEngineSettings.AllowAllUnknownUrlSchemes,
+        }
+    except Exception:
```
```suggestion
    except AttributeError:
```

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -162,6 +182,30 @@ class WebEngineSettings(websettings.AbstractSettings):
+
+    def _update_setting(self, setting, value):
+        if setting == 'content.unknown_url_scheme_policy':
+            if hasattr(self, '_UNKNOWN_URL_SCHEME_POLICY'):
```
Either having an attribute or not sounds a bit odd here - maybe set it to `None` instead and check that here? Something similar is done in other places for e.g. optional imports.

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -370,6 +370,23 @@ content.default_encoding:
+    valid_values:
+      - DisallowUnknownUrlSchemes: Disallows all navigation requests to URLs with unknown schemes.
+      - AllowUnknownUrlSchemesFromUserInteraction: Allows navigation requests to URLs with unknown schemes that are issued from user-interaction (like a mouse-click), whereas other navigation requests (for example from JavaScript) are suppressed.
+      - AllowAllUnknownUrlSchemes: Allows all navigation requests to URLs with unknown schemes.
```
Please use `disallow-unknown-url-schemes` etc. for consistency with other options.

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -370,6 +370,23 @@ content.default_encoding:
+  desc: >-
+    How navigation requests to URLs with unknown schemes are handled.
+
+    Only valid for webengine backend (Qt > 5.10). See
```
This shouldn't be needed anymore with the `backend` key (see the autogenerated text above).

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -370,6 +370,23 @@ content.default_encoding:
+    name: String
+    valid_values:
+      - DisallowUnknownUrlSchemes: Disallows all navigation requests to URLs with unknown schemes.
+      - AllowUnknownUrlSchemesFromUserInteraction: Allows navigation requests to URLs with unknown schemes that are issued from user-interaction (like a mouse-click), whereas other navigation requests (for example from JavaScript) are suppressed.
```
Please wrap this long line (see e.g. `content.cookies.accept` for an example)

The-Compiler commented on qutebrowser/config/configdata.yml:
```diff
@@ -370,6 +370,23 @@ content.default_encoding:
+    How navigation requests to URLs with unknown schemes are handled.
+
+    Only valid for webengine backend (Qt > 5.10). See
+    https://doc.qt.io/qt-5/qwebenginesettings.html#UnknownUrlSchemePolicy-enum.
```
I don't think this adds any value, as everything is now explained in the qutebrowser documentation anyways.

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -162,6 +182,30 @@ class WebEngineSettings(websettings.AbstractSettings):
         QWebEngineSettings.FantasyFont: QFont.Fantasy,
     }
 
+
```
One blank line only here (as per flake8).

The-Compiler commented:
Looks good now! The remaining issues on CI aren't your fault (a flaky test, and something I fixed in master already).

The only thing I'm left wondering is whether it's possible to set `supports_pattern: true` for the setting, i.e. if it's possible to do `:set -u example.com content.unknown_url_scheme_policy disallow-unknown-url-schemes` and only have it apply to that page. Could you try if that works (with two different sites with a magnet: link or so, for example)?

coiby commented:
On Wed, Apr 15, 2020 at 11:50:49PM -0700, Florian Bruhin wrote:
>Looks good now! The remaining issues on CI aren't your fault (a flaky test, and something I fixed in master already).
>
>The only thing I'm left wondering is whether it's possible to set `supports_pattern: true` for the setting, i.e. if it's possible to do `:set -u example.com content.unknown_url_scheme_policy disallow-unknown-url-schemes` and only have it apply to that page. Could you try if that works (with two different sites with a magnet: link or so, for example)?

Thank you for fixing the issues on CI!

I experienced the following problem after setting `supports_pattern:true`,

==== Exception ====
Traceback (most recent call last):
   File "qutebrowser/browser/webengine/webenginetab.py", line 1564, in _on_navigation_request
     changed = self.settings.update_for_url(navigation.url)
   File "qutebrowser/config/websettings.py", line 217, in update_for_url
     changed = self._update_setting(values.opt.name, value)
   File "qutebrowser/browser/webengine/webenginesettings.py", line 204, in _update_setting
     return self.set_unknown_url_scheme_policy(value)
   File "qutebrowser/browser/webengine/webenginesettings.py", line 195, in set_unknown_url_scheme_policy
     assert policy is not usertypes.UNSET  # type: ignore


One way to resolve this issue is to change AbstractSettings.update_for_url
so default value—allow-unknown-url-schemes-from-user-interaction in
this case—will be returned when calling get_for_url. This seems to make sense
because it could fix similar issues. But I'm not sure what other consequences
this change would introduce.

diff --git a/qutebrowser/config/websettings.py b/qutebrowser/config/websettings.py
index 03c85a71a..834874eb9 100644
--- a/qutebrowser/config/websettings.py
+++ b/qutebrowser/config/websettings.py
@@ -212,7 +212,7 @@ class AbstractSettings:
              if not values.opt.supports_pattern:
                  continue

-            value = values.get_for_url(url, fallback=False)
+            value = values.get_for_url(url, fallback=True)

              changed = self._update_setting(values.opt.name, value)
              if changed:

The-Compiler commented:
Nope, that should definitely stay at `UNSET`, because e.g. attributes should be reset with `resetAttribute` so that the global setting applies again (rather than being set to the same value than the global one, but still set).

Let's ignore that for now, then - I guess it makes sense to look at it when introducing URL pattern support for settings such was font size, font families and `default_encoding` in the future.

This LGTM now, except of a minor nitpick - I'll add a comment for that, but feel free to ignore it and I'll fix it up myself.

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -88,6 +88,13 @@ def fontFamily(self, which):
+    def unknownUrlSchemePolicy(self):
+        return self._settings[0].unknownUrlSchemePolicy()
+
+    def setUnknownUrlSchemePolicy(self, policy):
```
Nitpick: For consistency with the other methods/setters, this should be moved below `setDefaultTextEncoding` rather than here.

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -162,6 +182,30 @@ class WebEngineSettings(websettings.AbstractSettings):
+            True if there was a change, False otherwise.
+        """
+        if policy is usertypes.UNSET:
+            return False
```
I'm pretty sure this handling is incorrect - when navigating from a page with a customized URL scheme to a page without such a customization, nothing will be set now.

coiby replied:
I had some false assumptions before. After digging into the code, I realized this problem happened because after QWebEngineView is created, `WebEngineSettings._settings` will be changed from `_SettingsWrapper` to `QWebEngineSettings`. So `return False` when `policy is usertypes.UNNET` will not update the QWebEngineSettings,

```python
# qutebrowser/qutebrowser/browser/webengine/webenginetab.py
class WebEngineTab(browsertab.AbstractTab):
    def __init__(self, *, win_id, mode_manager, private, parent=None):
        ...
        widget = webview.WebEngineView(tabdata=self.data, win_id=win_id,
                                       private=private)
        ...
        # We're assigning settings in _set_widget
        self.settings = webenginesettings.WebEngineSettings(settings=None)
        self._set_widget(widget)

    def _set_widget(self, widget: QWidget) -> None:
        # pylint: disable=protected-access
        self._widget = widget
        ...
        self.settings._settings = widget.settings()
```

The-Compiler replied:
> After digging into the code, I realized this problem happened because after QWebEngineView is created, `WebEngineSettings._settings` will be changed from `_SettingsWrapper` to `QWebEngineSettings`.

I don't think so.

For settings specific to a tab (rather than global settings), `WebEngineSettings._settings` first is `None` (as set in the constructor); then as soon as the `QWebEngineView` is known, it's set to the `QWebEngineSettings` object.

For global settings, it's always the `_SettingsWrapper()`, in the global `webenginesettings.global_settings` instance of `WebEngineSettings`. The reason it exists is because there can be multiple profiles (default/private) which both have their own setting objects, but we want to apply all settings on both of them.

Then, depending on what changes, the code flow is different:

- For globally changed settings, `webenginesettings._update_settings` (the function) calls `global_settings.update_setting(...)` (which then calls `AbstractSettings._update_setting(...)` with the value, which then uses the `_SettingsWrapper`).
- For per-domain settings, the respective tab calls `AbstractSettings.update_for_url(...)`, which then calls `AbstractSettings._update_setting(...)`, which uses the underlying setting object directly (rather than the wrapper, because there is only one setting object it needs to change).

coiby replied:
Thank you for the detailed explaination! The reason why this problem happend is quite clear to me now.

coiby commented:
> Nope, that should definitely stay at `UNSET`, because e.g. attributes should be reset with `resetAttribute` so that the global setting applies again (rather than being set to the same value than the global one, but still set).
>
> Let's ignore that for now, then - I guess it makes sense to look at it when introducing URL pattern support for settings such was font size, font families and `default_encoding` in the future.
I see. I made some changes to handle UNSET in a different way in set_unknown_url_scheme_policy.

> The only thing I'm left wondering is whether it's possible to set `supports_pattern: true` for the setting, i.e. if it's possible to do `:set -u example.com content.unknown_url_scheme_policy disallow-unknown-url-schemes` and only have it apply to that page. Could you try if that works (with two different sites with a magnet: link or so, for example)?

`supports_pattern: true` is now supported. I've tested magnet: and tencen:// and it works as expected.


> This LGTM now, except of a minor nitpick - I'll add a comment for that, but feel free to ignore it and I'll fix it up myself.

Thank you for pointing out the inconsistency. I've fixed it in the last commits.

The-Compiler commented:
I'm confused by the last commit. Did you intend to move the code back to `websettings.py` (where it really doesn't belong, seeing that it uses QtWebEngine-specific API) and to go back from `_UNKNOWN_URL_SCHEME_POLICY = None` to `pass` (i.e. an attribute which is sometimes set, sometimes not)?

coiby commented:
> I'm confused by the last commit. Did you intend to move the code back to `websettings.py` (where it really doesn't belong, seeing that it uses QtWebEngine-specific API) and to go back from `_UNKNOWN_URL_SCHEME_POLICY = None` to `pass` (i.e. an attribute which is sometimes set, sometimes not)?

Yes, it's not not necessary to move the code back to support per-domain seeting. Thank you for 
correcting me!

The-Compiler commented on qutebrowser/browser/webengine/webenginesettings.py:
```diff
@@ -162,6 +182,32 @@ class WebEngineSettings(websettings.AbstractSettings):
+            True if there was a change, False otherwise.
+        """
+        old_value = self._settings.unknownUrlSchemePolicy()
+        if policy is usertypes.UNSET:  # type: ignore
```
This shouldn't be ignored anymore, because now, `usertypes.UNSET` can be a valid, expected value for this function. Instead, the type hint should probably be changed to `typing.Union[str, usertypes.Unset]` (`UNSET` is an instance of the class `Unset`).

coiby replied:
Unforturnately `tox -e mypy` complainted,

> qutebrowser/browser/webengine/webenginesettings.py:192: error: invalid syntax  [syntax]
            if policy is usertypes.UNSET:  # type: typing.Union[str, usertypes.Unset]

For _"error: Non-overlapping identity check (left operand type: "str", right operand type: "Unset")_, the only solution I can find is `type: ignore`. Or shall we remove type checking for set_unknown_url_scheme_policy as like other functions in qutebrowser/browser/webengine/webenginesettings.py?

The-Compiler replied:
Sorry for not being clearer - I mean the type hint for the `policy` argument, as `policy` now can be either `UNSET` or a string.

coiby replied:
I'm not familiar with mypy yet. Thank you for the clarification!
Is the following change OK?
```diff
diff --git a/qutebrowser/browser/webengine/webenginesettings.py b/qutebrowser/browser/webengine/webenginesettings.py
index f06e71653..00f501caf 100644
--- a/qutebrowser/browser/webengine/webenginesettings.py
+++ b/qutebrowser/browser/webengine/webenginesettings.py
@@ -182,18 +182,21 @@ class WebEngineSettings(websettings.AbstractSettings):
         QWebEngineSettings.FantasyFont: QFont.Fantasy,
     }
 
-    def set_unknown_url_scheme_policy(self, policy: str) -> bool:
+    def set_unknown_url_scheme_policy(
+                self,
+                policy: typing.Union[str, usertypes.Unset]
+    ) -> bool:
         """Set the UnknownUrlSchemePolicy to use.
 
         Return:
             True if there was a change, False otherwise.
         """
         old_value = self._settings.unknownUrlSchemePolicy()
-        if policy is usertypes.UNSET:  # type: ignore
+        if policy is usertypes.UNSET:
             self._settings.resetUnknownUrlSchemePolicy()
             new_value = self._settings.unknownUrlSchemePolicy()
         else:
-            new_value = self._UNKNOWN_URL_SCHEME_POLICY[policy]
+            new_value = self._UNKNOWN_URL_SCHEME_POLICY[policy]  # type: ignore
             self._settings.setUnknownUrlSchemePolicy(new_value)
         return old_value != new_value
 
```

The-Compiler replied:
I think if you replace `if policy is usertypes.UNSET` with `if isinstance(policy, usertypes.Unset)`, mypy should understand the code flow and you can drop the second `# type: ignore` as well.

coiby replied:
Thank you! That is a subtle change although I expected mypy to be smarter.

The-Compiler replied:
> That is a subtle change although I expected mypy to be smarter.

I'm not surprised it doesn't catch that - it doesn't have a way to know that `UNSET` is the only instance of `Unset`, i.e. that `... is UNSET` is equivalent to `isinstance(..., Unset)`. There *might* be a way for us to define `UNSET` in a way that it can know about this, but I'm not sure. I'll add some ideas to #5259.

coiby commented:
> > One way to resolve this issue is to change AbstractSettings.update_for_url
so default value—allow-unknown-url-schemes-from-user-interaction in
this case—will be returned when calling get_for_url. This seems to make sense
because it could fix similar issues. But I'm not sure what other consequences
this change would introduce.
>>```patch
>>diff --git a/qutebrowser/config/websettings.py b/qutebrowser/config/websettings.py
>>index 03c85a71a..834874eb9 100644
>>--- a/qutebrowser/config/websettings.py
>>+++ b/qutebrowser/config/websettings.py
>>@@ -212,7 +212,7 @@ class AbstractSettings:
>>              if not values.opt.supports_pattern:
>>                  continue
>>
>>-            value = values.get_for_url(url, fallback=False)
>>+            value = values.get_for_url(url, fallback=True)
>>
>>              changed = self._update_setting(values.opt.name, value)
>>              if changed:
>>```
>
> Nope, that should definitely stay at `UNSET`, because e.g. attributes should be reset with `resetAttribute` so that the global setting applies again (rather than being set to the same value than the global one, but still set).

While studing the quterbrowser code today, I rethink this idea.

If `resetAttribute` is equivalent to `setAttribute(attribute, default_value)`, we needn't to deal with the case usertypes.UNSET. Then the code can be simplified.

Accroding to the [QWebEngineSettings Class | Qt WebEngine 5.14.2](https://doc.qt.io/qt-5/qwebenginesettings.html#resetAttribute), this assumption seems to be true,
> void QWebEngineSettings::resetAttribute(QWebEngineSettings::WebAttribute attribute)
> Resets the setting of attribute to the value specified in the profile that the page belongs to.

> void QWebEngineSettings::resetFontFamily(QWebEngineSettings::FontFamily which)
Resets the actual font family specified by which to the one specified in the profile that the page belongs to.

> void QWebEngineSettings::resetFontSize(QWebEngineSettings::FontSize type)
Resets the font size for type to the size specified in the profile that the page belongs to.

Btw, I also find all values in config.instance which support pattern have default values.

The-Compiler commented:
> If `resetAttribute` is equivalent to `setAttribute(attribute, default_value)`, we needn't to deal with the case usertypes.UNSET. Then the code can be simplified.

There are other places in qutebrowser where this distinction needs to be made - there's a difference between "there is a config value set, and it happens to be the same as the default" and "there is no value set", and that distinction should stay.

I'm also pretty sure there is a distinction in QtWebEngine's C++ code as well, from what I remember from last time I read it. I also vaguely remember there being some issues some years back when qutebrowser did things that way.

> Btw, I also find all values in config.instance which support pattern have default values.

I don't understand what you mean with this.

coiby commented:
> > If `resetAttribute` is equivalent to `setAttribute(attribute, default_value)`, we needn't to deal with the case usertypes.UNSET. Then the code can be simplified.
> 
> There are other places in qutebrowser where this distinction needs to be made - there's a difference between "there is a config value set, and it happens to be the same as the default" and "there is no value set", and that distinction should stay.
>
> I'm also pretty sure there is a distinction in QtWebEngine's C++ code as well, from what I remember from last time I read it. I also vaguely remember there being some issues some years back when qutebrowser did things that way.

Then we should stick to the current approach. Thank you for clarifying this issue for me.

> > Btw, I also find all values in config.instance which support pattern have default values.
> 
> I don't understand what you mean with this.

I jsut wanted to make sure all the values supporting pattern matching have default value, i.e., when fallback=True, a default values will always be returned thus eliminating the need to deal with theusertypes.UNSET case.

The-Compiler commented:
Thanks! I decided to shorten the option values a bit:

- 728c84158 Shorten option values for content.unknown_url_scheme_policy
- d318e0d79 Fix unknown_url_scheme_policy default value

This is because Qt needs to repeat `unknown-url-schemes` in the enum members as those enums aren't namespaced/scoped - but we don't need to do that because they only exist in the context of that setting, so why not simplify them a bit. Thought of it last-minute after merging, so I did it myself rather than going for another round-trip :wink:

Thank you!

coiby commented:
On Mon, Apr 20, 2020 at 09:28:22AM -0700, Florian Bruhin wrote:
>Thanks! I decided to shorten the option values a bit:
>
>- 728c84158 Shorten option values for content.unknown_url_scheme_policy
>- d318e0d79 Fix unknown_url_scheme_policy default value
>
>This is because Qt needs to repeat `unknown-url-schemes` in the enum members as those enums aren't namespaced/scoped - but we don't need to do that because they only exist in the context of that setting, so why not simplify them a bit. Thought of it last-minute after merging, so I did it myself rather than going for another round-trip :wink:
>
>Thank you!

Great! And thank you for reviewing my work and all the detailed
feedbacks to help me better understand how qutebrowser works!
